lxd/identity/type.go

```diff
@@ -0,0 +1,247 @@
+		return fmt.Errorf("Unknown identity type %q", t)
+	}
+
+	return nil
+}
+
+// AuthenticationMethod returns the authentication method for the identity type. If the identity type is unknown, it returns an error.
+func (t Type) AuthenticationMethod() (string, error) {
+	err := t.Validate()
+	if err != nil {
+		return "", err
+	}
+
+	return identityTypes[t].authenticationMethod(), nil
+}
+
+// IsAdmin determines whether the identity type is an admin (unrestricted) identity. If the identity type is unknown, it returns an error.
+func (t Type) IsAdmin() (bool, error) {
+	err := t.Validate()
+	if err != nil {
+		return false, err
+	}
+
+	return identityTypes[t].isAdmin(), nil
+}
+
+// IsCacheable determines whether the identity type can be cached. If the identity type is unknown, it returns an error.
+func (t Type) IsCacheable() (bool, error) {
+	err := t.Validate()
+	if err != nil {
+		return false, err
+	}
+
+	return identityTypes[t].isCacheable(), nil
+}
+
+// IsFineGrained determines whether the identity type has fine-grained permissions (i.e. permissions are managed via group membership). If the identity type is unknown, it returns an error.
+func (t Type) IsFineGrained() (bool, error) {
+	err := t.Validate()
+	if err != nil {
+		return false, err
+	}
+
+	return identityTypes[t].isFineGrained(), nil
+}
+
+// typeInfo represents common attributes an identity type must have.
+type typeInfo interface {
+	authenticationMethod() string
+	isAdmin() bool
+	isCacheable() bool
+	isFineGrained() bool
+}
+
+var identityTypes = map[Type]typeInfo{
+	TypeOIDCClient:                     identityTypeOIDCClient{},
+	TypeCertificateClient:              identityTypeCertificateClient{},
+	TypeCertificateClientPending:       identityTypeCertificateClientPending{},
+	TypeCertificateClientRestricted:    identityTypeCertificateClientRestricted{},
+	TypeCertificateClientUnrestricted:  identityTypeCertificateClientUnrestricted{},
+	TypeCertificateMetricsRestricted:   identityTypeCertificateMetricsRestricted{},
+	TypeCertificateMetricsUnrestricted: identityTypeCertificateMetricsUnrestricted{},
+	TypeCertificateServer:              identityTypeCertificateServer{},
+}
+
+// typeInfoCommon partially implements [typeInfo] and can be embedded in [typeInfo] implementations for convenience.
+type typeInfoCommon struct{}
+
+// isAdmin returns false by default.
+func (typeInfoCommon) isAdmin() bool {
+	return false
+}
+
+// isCacheable returns false by default.
+func (typeInfoCommon) isCacheable() bool {
+	return false
+}
+
+// isFineGrained returns false by default.
+func (typeInfoCommon) isFineGrained() bool {
+	return false
+}
+
+type identityTypeOIDCClient struct {
+	typeInfoCommon
+}
+
+func (identityTypeOIDCClient) authenticationMethod() string {
+	return api.AuthenticationMethodOIDC
+}
+
+func (identityTypeOIDCClient) isCacheable() bool {
+	return true
+}
+
+func (identityTypeOIDCClient) isFineGrained() bool {
+	return true
+}
+
+type identityTypeCertificateClient struct {
```

Comment:
Please can you separate each identity type type definition by (at least) a title comment explaining what the identity type is for or perhaps into separate files, as otherwise the reviewer is faced with a long block of type definitions with no clear separation.